fix: network transform respawn tests fail on mac

NoelStephensUnity opened the pull request:
Determined there were timing issues between MAC and other platforms.
Also determined that MultiInstance timing varies from actual runtime local testing (manual testing).
Fixed the issue by refactoring the original fix to handle recycled (pooled) NetworkObjects.

### PR Checklist
- [X] Have you added a backport label (if needed)? For example, the `type:backport-release-*` label. After you backport the PR, the label changes to `stat:backported-release-*`.

## Testing and Documentation
* Includes refactoring of integration test .
* No documentation changes or additions were necessary.

NoelStephensUnity committed:
fix

Switching over to wait for frames as opposed to time in order to avoid timing issue with Mac (it runs slow).

NoelStephensUnity committed:
refactor

Determined I needed to approach this fix differently.
Refactored the NetworkTransform update and the test to verify the fix.

NoelStephensUnity committed:
Merge branch 'develop' into fix/NetworkTransformRespawnTests-fail-on-mac

NoelStephensUnity committed:
test refactor

Working out kinks in timing.

jeffreyrainy commented:
It feels like we are adding complexity to the NetworkTransform in order to address an issue where the initial position is not carried alongside the spawn. I wonder if it is the correct approach.

I would rather embrace the mismatch in position at spawn time and not care about it until we have proper synchronization between spawning and value update. This would leave us with a leaner and cleaner NetworkTransform. Where does the need to address this now, in this fashion, arise ?

NoelStephensUnity commented:
> It feels like we are adding complexity to the NetworkTransform in order to address an issue where the initial position is not carried alongside the spawn. I wonder if it is the correct approach.
> 
> I would rather embrace the mismatch in position at spawn time and not care about it until we have proper synchronization between spawning and value update. This would leave us with a leaner and cleaner NetworkTransform. Where does the need to address this now, in this fashion, arise ?

The primary issue is the testing of the fix and not the fix itself.

Since clients can't set NetworkVariables (even when the associated NetworkObject is not spawned), for recylced/re-used NetworkObjects we have to assign the m_ReplicatedNetworkState.Value to the m_LocalAuthoritativeNetworkState and then invoke the Initialization method only upon receiving the first state otherwise it will initialize with the NetworkObject's previously despawned m_ReplicatedNetworkState value.

NoelStephensUnity committed:
test fix

jeffreyrainy commented:
> > It feels like we are adding complexity to the NetworkTransform in order to address an issue where the initial position is not carried alongside the spawn. I wonder if it is the correct approach.
> > I would rather embrace the mismatch in position at spawn time and not care about it until we have proper synchronization between spawning and value update. This would leave us with a leaner and cleaner NetworkTransform. Where does the need to address this now, in this fashion, arise ?
> 
> The primary issue is the testing of the fix and not the fix itself.
> 
> Since clients can't set NetworkVariables (even when the associated NetworkObject is not spawned), for recylced/re-used NetworkObjects we have to assign the m_ReplicatedNetworkState.Value to the m_LocalAuthoritativeNetworkState and then invoke the Initialization method only upon receiving the first state otherwise it will initialize with the NetworkObject's previously despawned m_ReplicatedNetworkState value.

It is easier to review/understand if I look at just the NT and forget the test. With the above perspective it makes sense. I'll drop a comment or two, but please consider this resolved.

jeffreyrainy commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -652,7 +658,7 @@ private void AddInterpolatedState(NetworkTransformState newState)
         private void OnNetworkStateChanged(NetworkTransformState oldState, NetworkTransformState newState)
         {
-            if (!NetworkObject.IsSpawned)
+            if (!NetworkObject.IsSpawned || !gameObject.activeInHierarchy)
```
Where does the `gameObject.activeInHierarchy` bit comes from ? I don't see how it ties in with the rest.

NoelStephensUnity replied:
Still testing things...

NoelStephensUnity committed:
test fix

using a more direct approach to the per frame motion and made a few adjustments regarding frames waiting after spawn (etc).

NoelStephensUnity committed:
test

added debug log
